srtcore/core.cpp

```diff
@@ -14468,3 +14531,1224 @@ int CUDTGroup::configure(const char* str)
 
     return 0;
 }
+
+int CUDTGroup::baseOffset(SRT_MSGCTRL& mctrl)
+{
+    return CSeqNo::seqoff(m_RcvBaseSeqNo, mctrl.pktseq);
+}
+
+int CUDTGroup::baseOffset(CUDTGroup::ReadPos& pos)
+{
+    return CSeqNo::seqoff(m_RcvBaseSeqNo, pos.mctrl.pktseq);
+}
+
+bool CUDTGroup::seqDiscrepancy(SRT_MSGCTRL& mctrl)
+{
+    // Here it's predicted that the incoming packet's seq may only be
+    // at most 1/8 of the whole range in advance.
+
+    // seqoff returns the distance between two sequences,
+    // which is the less of the two possible distances that
+    // one can be reached from another. The precondition
+    // is here that this difference, in whichever direction,
+    // shall not be greater than 1/4 of the whole wheel.
+    return abs(CSeqNo::seqoff(m_RcvBaseSeqNo, mctrl.pktseq)) > CSeqNo::m_iSeqNoTH/2;
+}
+
+bool CUDTGroup::msgDiscrepancy(SRT_MSGCTRL& mctrl)
+{
+    // Here it's predicted that the incoming packet's seq may only be
+    // at most 1/8 of the whole range in advance.
+
+    // seqoff returns the distance between two sequences,
+    // which is the less of the two possible distances that
+    // one can be reached from another. The precondition
+    // is here that this difference, in whichever direction,
+    // shall not be greater than 1/4 of the whole wheel.
+    return abs(MsgNo(mctrl.msgno) - MsgNo(m_RcvBaseMsgNo)) > int32_t(MsgNo::HALF/2);
+}
+
+
+int CUDTGroup::recvBalancing(char* buf, int len, SRT_MSGCTRL& w_mctrl)
+{
+    typedef map<SRTSOCKET, ReadPos>::iterator pit_t;
+    // Later iteration over it might be less efficient than
+    // by vector, but we'll also often try to check a single id
+    // if it was ever seen broken, so that it's skipped.
+    set<CUDTSocket*> broken;
+
+    // Remember them now because they will be overwritten.
+    SRT_SOCKGROUPDATA* out_grpdata = (w_mctrl.grpdata);
+    size_t out_grpdata_size = w_mctrl.grpdata_size;
+
+    size_t output_size = 0;
+
+RETRY_READING:
```

Comment:
This goto label can be replaced with `while(true)` loop.

Reply:
Yeah, the problem is that it would likely embrace the whole function. This function is also quite long and would be nice if it can be split into smaller parts. With this goto form at least you have it in the beginning and in the end of the function and therefore it's clear enough: repeat everything from start.